### PR TITLE
[ReadMscwCtm] Supporting selecting files to be read by regexp match

### PR DESCRIPTION
**Pull request by thorbjoernl**
## Change Summary

Support providing a custom regular expression to the EMEP reader which can be used to select files instead of the Base_{freq}.nc` matching that is currently being used.

Currently [works](https://aeroval-test.met.no/thlun/pages/evaluation/?project=uemep_test&experiment=20241125) with concatenated files.

## Related issue number

- #1175 

## Checklist

* [x] Start with a draft-PR
* [x] The PR title is a good summary of the changes
* [x] PR is set to AeroTools and a tentative milestone
* [x] Documentation reflects the changes where applicable
* [x] Tests for the changes exist where applicable
* [x] Tests pass locally
* [x] Tests pass on CI
* [x] At least 1 reviewer is selected
* [x] Make PR ready to review

### CODE REVIEW

File: pyaerocom/io/mscw_ctm/reader.py [hunk 1]
```diff
@@ -194,8 +196,20 @@ def __init__(self, data_id=None, data_dir=None, **kwargs):
             if isinstance(new_map, dict):
                 self.var_map.update(new_map)
             else:
-                logger.warn(f"New map {new_map} is not a dict. Skipping")
+                logger.warning(f"New map {new_map} is not a dict. Skipping")
 
+        if (pattern := kwargs.get("file_pattern", None)) is None:
```
Comment (heikoklein) [inline pyaerocom/io/mscw_ctm/reader.py]: overuse of wallruss operator

File: pyaerocom/io/mscw_ctm/reader.py [hunk 2]
```diff
@@ -194,8 +196,20 @@ def __init__(self, data_id=None, data_dir=None, **kwargs):
...
+            # Pattern for the 'Base_{freq}.nc' default strategy.
+            pattern = rf"^Base_({'|'.join(self.FREQ_CODES.keys())}).nc$"
+
+        if not isinstance(pattern, str | re.Pattern):
+            raise TypeError(
+                f"Provided pattern '{pattern}' of type {type(pattern)} can't be compiled to re.Pattern. Please provide str or re.Pattern."
```
Comment (heikoklein) [inline pyaerocom/io/mscw_ctm/reader.py]: The user only knows `file_pattern`, not `pattern`. And you haven't tried to compile yet, did you?

File: pyaerocom/io/mscw_ctm/reader.py [hunk 3]
```diff
@@ -182,8 +183,9 @@ class _PrivateFields:
         filepaths: list[str] | None = None
         files: list[str] | None = None
         data_dir: str | None = None
+        file_pattern: re.Pattern
 
-    def __init__(self, data_id=None, data_dir=None, **kwargs):
+    def __init__(self, data_id: str | None = None, data_dir: str | None = None, **kwargs):
```
Comment (heikoklein) [inline pyaerocom/io/mscw_ctm/reader.py]: documentation of file_pattern missing. Why not raise it to usual parameters and hide it in kwargs?

File: pyaerocom/io/mscw_ctm/reader.py [hunk 4]
```diff
@@ -244,24 +258,27 @@ def _get_trend_folders_from_folder(self):
...
         for d in os.listdir(dd):
             if not os.path.isdir(os.path.join(dd, d)):
                 continue
             m = re.match(self.YEAR_PATTERN, d)
             if m is not None:
-                has_mscwfiles = False
-                for f in mscwfiles:
-                    if os.path.exists(os.path.join(dd, d, f)):
-                        has_mscwfiles = True
+                if self._private.file_pattern is not None:
```
Comment (heikoklein) [inline pyaerocom/io/mscw_ctm/reader.py]: how can that be None? It must be set in init.

**Review by heikoklein — CHANGES_REQUESTED** [review]

**Comment by thorbjoernl** [comment]
Closing because changes are merged with #1425

**Pull request closed without merging**